### PR TITLE
http: remove unnecessary http formatter interface and update HeaderEvaluator

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: http: remove unnecessary http formatter interface
Additional Description:

The header formatter basically is a historical remains. It is completely unnecessary today because substitution formatter will be used for both logging and HTTP header mutations.

Risk Level: low.
Testing: unit..
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

File: envoy/http/header_evaluator.h [hunk 1]
```diff
@@ -16,13 +17,11 @@ class HeaderEvaluator {
...
    * @param headers the target header map to be mutated.
-   * @param request_headers request headers to be used in the header manipulation.
-   * @param response_headers response headers to be used in the header manipulation.
+   * @param format_context context to format the header value.
    * @param stream_info the source of values that can be used in the evaluation.
    */
   virtual void evaluateHeaders(Http::HeaderMap& headers,
-                               const Http::RequestHeaderMap& request_headers,
-                               const Http::ResponseHeaderMap& response_headers,
+                               const Formatter::HttpFormatterContext& format_context,
```
Comment (wbpcode) [inline envoy/http/header_evaluator.h:24]: The http formatter interface is removed, so I also updated the HeaderEvaluator to use new `HttpFormatterContext`.

**Comment by wbpcode** [comment]
Hi, @cpakulski , could you take a first look when you have free time. I remember that you did lots of great job on this area.

**Comment by cpakulski** [comment]
Yes, of course! I will review it tomorrow morning.

File: source/common/router/config_impl.cc [hunk 2]
```diff
@@ -863,11 +863,8 @@ void RouteEntryImplBase::finalizeResponseHeaders(Http::ResponseHeaderMap& header
   for (const HeaderParser* header_parser : getResponseHeaderParsers(
            /*specificity_ascend=*/vhost_->globalRouteConfig().mostSpecificHeaderMutationsWins())) {
     // Later evaluated header parser wins.
-    header_parser->evaluateHeaders(headers,
-                                   stream_info.getRequestHeaders() == nullptr
-                                       ? *Http::StaticEmptyHeaders::get().request_headers
-                                       : *stream_info.getRequestHeaders(),
-                                   headers, stream_info);
+    header_parser->evaluateHeaders(headers, {stream_info.getRequestHeaders(), &headers},
```
Comment (cpakulski) [inline source/common/router/config_impl.cc:866]: What happens if `getRequestHeaders` returns nullptr? Would `StaticEmptyRequestHeaders` be used?

File: source/extensions/filters/http/header_mutation/header_mutation.cc [hunk 3]
```diff
@@ -42,26 +41,15 @@ Http::FilterHeadersStatus HeaderMutation::decodeHeaders(Http::RequestHeaderMap&
...
-  const Http::RequestHeaderMap* downstream_request_headers =
-      encoder_callbacks_->streamInfo().getRequestHeaders();
-  const Http::RequestHeaderMap& request_headers =
-      downstream_request_headers != nullptr
-          ? *downstream_request_headers
-          : *Http::StaticEmptyHeaders::get().request_headers.get();
-
-  config_->mutations().mutateResponseHeaders(request_headers, headers,
-                                             encoder_callbacks_->streamInfo());
+  Formatter::HttpFormatterContext ctx{encoder_callbacks_->requestHeaders().ptr(), &headers};
```
Comment (cpakulski) [inline source/extensions/filters/http/header_mutation/header_mutation.cc:51]: Should we worry about described corner case? I believe that if that situation happens, `encoder_callbacks_->requestHeaders().ptr()` will be nullptr and the context will handle it gracefully, correct?

File: test/common/router/header_formatter_speed_test.cc [hunk 4]
```diff
@@ -40,8 +40,7 @@ static void bmEvaluateHeaders(benchmark::State& state) {
   HeaderParserPtr header_parser = HeaderParser::configure(headers_to_add);
 
   for (auto _ : state) { // NOLINT: Silences warning about dead store
-    header_parser->evaluateHeaders(*request_header, *request_header,
-                                   *Http::StaticEmptyHeaders::get().response_headers, *stream_info);
+    header_parser->evaluateHeaders(*request_header, {request_header.get()}, *stream_info);
```
Comment (cpakulski) [inline test/common/router/header_formatter_speed_test.cc:43]: Was there any effect on the speed test?

**Review by cpakulski — COMMENTED** [review]
Code changes LGTM. Just few questions.

File: source/common/router/config_impl.cc (hunk 2 above)
Comment (wbpcode) [reply source/common/router/config_impl.cc:866]: Yeah, the `HttpFormatterContext` will handle it correctly.

File: source/extensions/filters/http/header_mutation/header_mutation.cc (hunk 3 above)
Comment (wbpcode) [reply source/extensions/filters/http/header_mutation/header_mutation.cc:51]: Yeah.

File: test/common/router/header_formatter_speed_test.cc (hunk 4 above)
Comment (wbpcode) [reply test/common/router/header_formatter_speed_test.cc:43]: I think there should have no obvious impact. But I can take a check and paste result here.

Comment (wbpcode) [reply test/common/router/header_formatter_speed_test.cc:43]: Result after this PR:
```
---------------------------------------------------------------
Benchmark                     Time             CPU   Iterations
---------------------------------------------------------------
bmEvaluateHeaders/2         316 ns          316 ns      2155852
bmEvaluateHeaders/4         982 ns          982 ns       715449
bmEvaluateHeaders/6        1564 ns         1564 ns       441221
bmEvaluateHeaders/8        2118 ns         2118 ns       336647
bmEvaluateHeaders/10       2701 ns         2701 ns       258921
bmEvaluateHeaders/12       3310 ns         3309 ns       212408
bmEvaluateHeaders/14       3943 ns         3943 ns       181631
bmEvaluateHeaders/16       4442 ns         4442 ns       156797
bmEvaluateHeaders/18       5178 ns         5178 ns       138296
bmEvaluateHeaders/20       5737 ns         5737 ns       120386
```

Comment (wbpcode) [reply test/common/router/header_formatter_speed_test.cc:43]: Result before this PR:
```
---------------------------------------------------------------
Benchmark                     Time             CPU   Iterations
---------------------------------------------------------------
bmEvaluateHeaders/2         341 ns          341 ns      2057969
bmEvaluateHeaders/4        1032 ns         1032 ns       671125
bmEvaluateHeaders/6        1670 ns         1670 ns       416722
bmEvaluateHeaders/8        2330 ns         2329 ns       301196
bmEvaluateHeaders/10       3018 ns         3017 ns       234001
bmEvaluateHeaders/12       3717 ns         3717 ns       191802
bmEvaluateHeaders/14       4269 ns         4269 ns       165998
bmEvaluateHeaders/16       4873 ns         4873 ns       143795
bmEvaluateHeaders/18       5529 ns         5528 ns       126884
bmEvaluateHeaders/20       6219 ns         6219 ns       113101
```

**Comment by wbpcode** [comment]
cc @cpakulski any new concern?

File: envoy/formatter/substitution_formatter.h [hunk 5]
```diff
@@ -100,9 +100,9 @@ class HttpFormatterContext {
   AccessLog::AccessLogType accessLogType() const;
 
 private:
-  const Http::RequestHeaderMap* request_headers_{};
-  const Http::ResponseHeaderMap* response_headers_{};
-  const Http::ResponseTrailerMap* response_trailers_{};
+  mutable const Http::RequestHeaderMap* request_headers_{};
+  mutable const Http::ResponseHeaderMap* response_headers_{};
+  mutable const Http::ResponseTrailerMap* response_trailers_{};
```
Comment (RyanTheOptimist) [inline envoy/formatter/substitution_formatter.h:105]: Are these changes necessary? I wouldn't think so since their private and we're not changing the const-ness of any methods (or adding any new const methods).

File: source/common/formatter/http_specific_formatter.cc [hunk 6]
```diff
@@ -29,16 +29,22 @@ HttpFormatterContext::HttpFormatterContext(const Http::RequestHeaderMap* request
       log_type_(log_type) {}
 
 const Http::RequestHeaderMap& HttpFormatterContext::requestHeaders() const {
-  return request_headers_ != nullptr ? *request_headers_
-                                     : *Http::StaticEmptyHeaders::get().request_headers;
+  if (request_headers_ == nullptr) {
+    request_headers_ = Http::StaticEmptyHeaders::get().request_headers.get();
+  }
+  return *request_headers_;
```
Comment (RyanTheOptimist) [inline source/common/formatter/http_specific_formatter.cc:35]: Are these change for any particular reason?

**Review by RyanTheOptimist — COMMENTED** [review]
Wow, nice. Thanks for doing this. One minor question.

File: envoy/formatter/substitution_formatter.h (hunk 5 above)
Comment (wbpcode) [reply envoy/formatter/substitution_formatter.h:105]: We may update them in the requestHeaders/responseHeaders.

File: source/common/formatter/http_specific_formatter.cc (hunk 6 above)
Comment (wbpcode) [reply source/common/formatter/http_specific_formatter.cc:35]: The requestHeaders/responseHeaders methods maybe called multiple times in same formatting, and this change could ensure the `StaticEmptyHeaders::get()` only be called once.

These empty headers are constructed on first use. So, the `StaticEmptyHeaders::get()` require at least one time atomic bool checking.
This change want to reduce these atomic operations.

This is pretty simple change and from the benchmark results, this change at least not make things worse.

(I actually have no insistence on this point. If you think this is unnecessary, I can revert it or add a new benchmark case to ensure it's effect.)

Comment (RyanTheOptimist) [reply source/common/formatter/http_specific_formatter.cc:35]: Ah, I see. I'm not an enormous fan of mutable members if we can avoid them. I'd be inclined to revert this part of the change and write a benchmark to see if it makes a difference, and then reassess. But up to you.

**Review by RyanTheOptimist — DISMISSED** [review]
Please update the PR description to explain the changes to evaluateHeaders.

**Comment by wbpcode** [comment]
cc @RyanTheOptimist could you give a new stamp again? thanks.

**Review by RyanTheOptimist — APPROVED** [review]
